lib/ReactViews/Story/StoryBuilder.jsx

```diff
@@ -0,0 +1,160 @@
+import React from 'react';
+import createReactClass from 'create-react-class';
+import PropTypes from 'prop-types';
+import combine from 'terriajs-cesium/Source/Core/combine';
+import {getShareData } from '../Map/Panels/SharePanel/BuildShareLink';
+import defined from 'terriajs-cesium/Source/Core/defined';
+import ObserveModelMixin from '../ObserveModelMixin';
+import Icon from "../Icon.jsx";
+import Story from './Story.jsx';
+import StoryEditor from './StoryEditor.jsx';
+import Sortable from 'react-anything-sortable';
+import uniqid from 'uniqid';
+
+import Styles from './story-builder.scss';
+import '!!style-loader!css-loader?sourceMap!react-anything-sortable/sortable.css';
+
+function arrayMove(arr, oldIndex, newIndex) {
+    if (newIndex >= arr.length) {
+        let k = newIndex - arr.length + 1;
+        while (k--) {
+            arr.push(undefined);
+        }
+    }
+    arr.splice(newIndex, 0, arr.splice(oldIndex, 1)[0]);
+    return arr; 
+}
+
+const StoryBuilder = createReactClass({
+    displayName: 'StoryBuilder',
+    mixins: [ObserveModelMixin],
+    propTypes: {
+        terria: PropTypes.object.isRequired,
+        viewState: PropTypes.object.isRequired
+    },
+
+    getInitialState() {
+        return {
+            editingMode: false,
+            currentStory: undefined
+        };
+    },
+
+    removeStory(index, story) {
+      this.props.terria.stories = this.props.terria.stories.filter(st => st.id !== story.id);
+      if(index < this.props.viewState.currentStoryId) {
+        this.props.viewState.currentStoryId -= 1;
+      }
+    },
+    onSave(_story) {
+      const story = {
+        title: _story.title,
+        text: _story.text,
+        id: _story.id ? _story.id : uniqid(),
+      };
+
+      const storyIndex = (this.props.terria.stories || []).map(story => story.id).indexOf(_story.id);
+    
+      if(storyIndex >= 0) {
+        const oldStory = this.props.terria.stories[storyIndex];
+        // replace the old story, we need to replace the stories array so that
+        // it is observable
+        this.props.terria.stories = [...this.props.terria.stories.slice(0, storyIndex), combine(story, oldStory), ...this.props.terria.stories.slice(storyIndex + 1)];
+      } else {
+        this.captureStory(story);
+      }
+
+      this.setState({
+          editingMode: false
+      });
+  },
+
+    captureStory(story) {
+        story.shareData = JSON.parse(JSON.stringify(getShareData(this.props.terria, false)));
+        this.props.terria.stories = [...(this.props.terria.stories || []), story];
```

Comment:
ok `push` works but i'm confused here, since `stories` is tracked by knockout, i thought it needs to be replaced (rather than modified) for knockout to notice the change?